crates/autopilot/src/domain/settlement/trade.rs

```diff
@@ -39,102 +43,491 @@ impl Trade {
+        // So:
+        //    factor' = fee / surplus_after_fee = fee / (surplus_before_fee -
+        // fee) = fee / ((fee / factor) - fee) = factor / (1 - factor)
+        //
+        // Finally:
+        //     fee = surplus_after_fee * factor / (1 - factor)
+        let fee = surplus
+            .amount
+            .apply_factor(factor / (1.0 - factor))
+            .ok_or(error::Math::Overflow)?;
+
+        Ok(eth::Asset {
+            token: surplus.token,
+            amount: fee,
+        })
+    }
+
+    /// Protocol fee as a cut of the trade volume, denominated in SURPLUS token
+    fn volume_fee(&self, factor: f64) -> Result<eth::Asset, Error> {
+        // Volume fee is specified as a `factor` from raw volume (before fee). Since
+        // this module works with trades that already have the protocol fee applied, we
+        // need to calculate the protocol fee as an observation of a the eventually
+        // traded amount using a different factor `factor'` .
+        //
+        // The protocol fee before being applied is:
+        // case Sell: fee = traded_buy_amount * factor, resulting in the REDUCED
+        // buy amount
+        // case Buy: fee = traded_sell_amount * factor, resulting in the INCREASED
+        // sell amount
+        //
+        // The protocol fee after being applied is:
+        // case Sell: fee = traded_buy_amount' * factor',
+        // case Buy: fee = traded_sell_amount' * factor',
+        //
+        // Also:
+        // case Sell: traded_buy_amount' = traded_buy_amount - fee
+        // case Buy: traded_sell_amount' = traded_sell_amount + fee
+        //
+        // So:
+        // case Sell: factor' = fee / (traded_buy_amount - fee) = fee / (fee /
+        // factor - fee) = factor / (1 - factor)
+        // case Buy: factor' = fee / (traded_sell_amount + fee) = fee / (fee /
+        // factor + fee) = factor / (1 + factor)
+        //
+        // Finally:
+        // case Sell: fee = traded_buy_amount' * factor / (1 - factor)
+        // case Buy: fee = traded_sell_amount' * factor / (1 + factor)
+        let executed_in_surplus_token: eth::TokenAmount = match self.side {
+            order::Side::Sell => self
+                .executed
+                .0
+                .checked_mul(self.prices.custom.sell)
+                .ok_or(error::Math::Overflow)?
+                .checked_div(self.prices.custom.buy)
+                .ok_or(error::Math::DivisionByZero)?,
+            order::Side::Buy => self
+                .executed
+                .0
+                .checked_mul(self.prices.custom.buy)
+                .ok_or(error::Math::Overflow)?
+                .checked_div(self.prices.custom.sell)
+                .ok_or(error::Math::DivisionByZero)?,
+        }
+        .into();
+        let factor = match self.side {
+            order::Side::Sell => factor / (1.0 - factor),
+            order::Side::Buy => factor / (1.0 + factor),
```

Comment:
I know this isn't possible in the current code path, but should we be defensive here and return math errors in case we divide by 0 (we just had a panic in prod this week)

Reply:
it would result in `f64::NaN` and be handled without panic. https://users.rust-lang.org/t/how-should-i-handle-f64-division-by-zero-panic-elegantly/32485/2

Unless you are worried we might in the future add some more code right after this line which might have assumptions that f64 is not nan?

Reply:
In this case `apply_factor` will return 0 if I'm not mistake (`from_f64_lossy` for NaN yields 0), which is wrong (volume fee should be large in that case not 0) and going to be gnarly to debug. Why can't we simply return a `Math::DivisionByZero` error in this case (which intuitively is the correct thing to do here)?